underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java

```diff
@@ -0,0 +1,693 @@
+      mClient.getObjectMetadata(mBucketName, keyAsFolder);
+      // If no exception is thrown, the key exists as a folder
+      return true;
+    } catch (AmazonClientException e) {
+      // It is possible that the folder has not been encoded as a _$folder$ file
+      try {
+        String dir = stripPrefixIfPresent(key);
+        String dirPrefix = dir.endsWith(PATH_SEPARATOR) ? dir : dir + PATH_SEPARATOR;
+        // Check if anything begins with <folder_path>/
+        ObjectListing objs = mClient.listObjects(mBucketName, dirPrefix);
+        // If there are, this is a folder and we can create the necessary metadata
+        if (objs.getObjectSummaries().size() > 0) {
+          mkdirsInternal(dir);
+          return true;
+        } else {
+          return false;
+        }
+      } catch (AmazonClientException ace) {
+        return false;
+      }
+    }
+  }
+
+  /**
+   * Checks if the key is the root.
+   *
+   * @param key the key to check
+   * @return true if the key is the root, false otherwise
+   */
+  private boolean isRoot(String key) {
+    return PathUtils.normalizePath(key, PATH_SEPARATOR).equals(
+        PathUtils.normalizePath(Constants.HEADER_S3A + mBucketName, PATH_SEPARATOR));
+  }
+
+  /**
+   * Lists the files in the given path, the paths will be their logical names and not contain the
+   * folder suffix. Note that, the list results are unsorted.
+   *
+   * @param path the key to list
+   * @param recursive if true will list children directories as well
+   * @return an array of the file and folder names in this directory
+   * @throws IOException if an I/O error occurs
+   */
+  private String[] listInternal(String path, boolean recursive) throws IOException {
+    path = stripPrefixIfPresent(path);
+    path = PathUtils.normalizePath(path, PATH_SEPARATOR);
+    path = path.equals(PATH_SEPARATOR) ? "" : path;
+    String delimiter = recursive ? "" : PATH_SEPARATOR;
+    Set<String> children = new HashSet<>();
+    try {
+      ListObjectsV2Request request =
+          new ListObjectsV2Request().withBucketName(mBucketName).withPrefix(path)
+              .withDelimiter(delimiter).withMaxKeys(LISTING_LENGTH);
+      ListObjectsV2Result result = null;
+      while (result == null || result.isTruncated()) {
+        // Query S3 for the next batch of objects
+        result = mClient.listObjectsV2(request);
+        // Advance the request continuation token to the next set of objects
+        request.setContinuationToken(result.getNextContinuationToken());
+
+        // Directories in S3 UFS can be possibly encoded in two different ways:
+        // (1) as file objects with FOLDER_SUFFIX for directories created through Alluxio or
+        // (2) as "common prefixes" of other files objects for directories not created through
+        // Alluxio
+        //
+        // Case (1) (and file objects) is accounted for by iterating over chunk.getObjects() while
```

Comment:
Are you referring to the case:
```
/foo/bar
/foo_$folder$?
```
In this case we will pick up dir `/foo` from `/foo_$folder$` and ignore the common prefix (we check contains, though its technically unnecessary since it is a set)